optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/common/ConstraintCollectors.java

```diff
@@ -0,0 +1,194 @@
+                    resultContainer[0]++;
+                    return (() -> resultContainer[0]--);
+                },
+                resultContainer -> resultContainer[0]);
+    }
+
+    public static <A, B> BiConstraintCollector<A, B, ?, Integer> countBi() {
+        return new BiConstraintCollector<>(
+                () -> new int[1],
+                (resultContainer, a, b) -> {
+                    resultContainer[0]++;
+                    return (() -> resultContainer[0]--);
+                },
+                resultContainer -> resultContainer[0]);
+    }
+
+    public static <A, B> BiConstraintCollector<A, B, ?, Long> countLongBi() {
+        return new BiConstraintCollector<>(
+                () -> new long[1],
+                (resultContainer, a, b) -> {
+                    resultContainer[0]++;
+                    return (() -> resultContainer[0]--);
+                },
+                resultContainer -> resultContainer[0]);
+    }
+
+    // ************************************************************************
+    // countDistinct
+    // ************************************************************************
+
+    public static <A> UniConstraintCollector<A, ?, Integer> countDistinct(Function<A, ?> groupValueMapping) {
+        class CountDistinctResultContainer {
+            int count = 0;
+            Map<Object, int[]> objectCountMap = new HashMap<>();
+        }
+        return new UniConstraintCollector<>(
+                CountDistinctResultContainer::new,
+                (resultContainer, a) -> {
+                    Object value = groupValueMapping.apply(a);
+                    int[] objectCount = resultContainer.objectCountMap.computeIfAbsent(value, k -> new int[1]);
+                    if (objectCount[0] == 0) {
+                        resultContainer.count++;
+                    }
+                    objectCount[0]++;
+                    return (() -> {
+                        int[] objectCount2 = resultContainer.objectCountMap.get(value);
+                        if (objectCount2 == null) {
+                            throw new IllegalStateException("Impossible state: the value (" + value
+                                    + ") of A (" + a + ") is removed more times than it was added.");
+                        }
+                        objectCount2[0]--;
+                        if (objectCount2[0] == 0) {
+                            resultContainer.objectCountMap.remove(value);
+                            resultContainer.count--;
+                        }
+                    });
+                },
+                resultContainer -> resultContainer.count);
+    }
+
+    public static <A> UniConstraintCollector<A, ?, Long> countDistinctLong(Function<A, ?> groupValueMapping) {
+        class CountDistinctResultContainer {
+            long count = 0L;
+            Map<Object, long[]> objectCountMap = new HashMap<>();
+        }
+        return new UniConstraintCollector<>(
```

Comment:
A constructor call, that passes there a method reference, a lambda containing a few more lambdas and as a last argument, another lambda, this time a simple one. This really does not read well, does it?

Reply:
No. It's a poor excuse, but this is from OpenJDK:

    static class CollectorImpl<T, A, R> implements Collector<T, A, R> {
        private final Supplier<A> supplier;
        private final BiConsumer<A, T> accumulator;
        private final BinaryOperator<A> combiner;
        private final Function<A, R> finisher;
        private final Set<Characteristics> characteristics;

        CollectorImpl(Supplier<A> supplier,
                      BiConsumer<A, T> accumulator,
                      BinaryOperator<A> combiner,
                      Function<A,R> finisher,
                      Set<Characteristics> characteristics) {
            this.supplier = supplier;
            this.accumulator = accumulator;
            this.combiner = combiner;
            this.finisher = finisher;
            this.characteristics = characteristics;
        }

How can we improve our naming?